### DISCUSSION TITLE
how to use "useEffect", when useRouter() is undefined  in the first render?

### DISCUSSION THREAD

**Discussion by LuYujin9** [neuefische/questions#556] [Web] [answered]
The useState in my code can't work. I found the reason, but I can't fix it. Could you please help me?

1. my useRouter() can't return a value in the first render, so I use `if (!router.isReady) return;` to check it.
2. to avoid too many re-render I should use "useEffect", but I can't use "useEffect" after `if (!router.isReady) return;`, but without it, I can't access the data. 

When click the link to page "/restaurants/menu/1", there is no error. But when the page "/restaurants/menu/1" is refreshed, the error shows there.

my problem is in Pages/restaurants/menu/[id].js
here is my code. 
https://github.com/LuYujin9/Problem-router.isready-useEffect/blob/one-solution/pages/restaurants/menu/%5Bid%5D.js

**Comment by Gblur** [comment]
Where do you setIndex?

**Reply by LuYujin9** [reply]
A button with function

```javascript
 function handleToNextPage() {
    index < foods.length - 5 ? setIndex(index + 5) : null;
  }
```
```javascript
 <button aria-label="to next page" onClick={handleToNextPage}>
            Next Page
          </button>
```

**Comment by mpagels** [comment]
Hey @LuYujin9 I can't understand the problem. The code is running on my machine and I see no re-renders. Can you provide more information about how to reproduce your problem.

**Reply by LuYujin9** [reply]
Thanks Martin. the error shows there without useEffect. so I use "useEffect" to fixed it.  But there is a red line under "useEffect" ,  it shows "React Hook "useEffect" is called conditionally. React Hooks must be called in the exact same order in every component render."

**Reply by LuYujin9** [reply]
Florian checks the code with me. He noticed that, the reason can be from here:
```javascript
 const router = useRouter();
  if (!router.isReady) {
    return <h2>Loading...</h2>;
  }
  const { id } = router.query;
````
React is throwing the above error because we are calling [useEffect]after the return statement (inside isReady check).

Problem ist, without this code, my ganz website can be rendered.

**Reply by mpagels** [reply]
move this code

```js
  if (!router.isReady) {
    return <h2>Loading...</h2>;
  }
```

just above the other return

snippet:
```js
console.log(foods.slice(index, index + 5));
console.log(index);

if (!router.isReady) {
  return <h2>Loading...</h2>;
}

  return (
    <>
      <Heading previousLevelUrl={`/restaurants/${id}`}>
        {restaurant.name}
      </Heading>
      <StyledSection>
        <StyledImage
```

**Reply by LuYujin9** [reply]
Thanks @mpagels Martin. We have tried this way, but I also doesn't work.
Normally the "id" will be accessed with this code ` const router = useRouter(); const { id } = router.query;`, But it's weird, it can't work in my code.
It seems like useRouter runs slow, "id" will be "undefined". My page can't render.  that's the reason why I check "isReady" before `const { id } = router.query;`

**Reply by Gblur** [reply]
think I found a solution

**Reply by Gblur** [reply]
```  const [index, setIndex] = useState(0);
  const [restaurant, setRestaurant] = useState({
    id: "",
    name: "",
    price: "",
    description: ""
  })
  const [FoodsInCurrentPage, setFoodsInCurrentPage] = useState([
    { name: "Noch keine Speisekarte" },
  ]);


  const router = useRouter();
 
  const { id } = router.query;

  function getRestaurant() {  
      return restaurants.find((restaurant) => restaurant.id === id);
  }

  //Problem is here.
  useEffect(() => {
    if(id){
      setRestaurant(getRestaurant()) 
    }
    if(restaurant.foods){
      setFoodsInCurrentPage(restaurant.foods.slice(index, index + 5));
    }
    return () => {

    }
  }, [setRestaurant, restaurant.foods, id]);

  console.log(restaurant.foods)
  
  function handleToNextPage() {
    index < FoodsInCurrentPage.length - 5 ? setIndex(index + 5) : null;
  }
  function handleToLastPage() {
    index > 4 ? setIndex(index - 5) : null;
  }

  if (!router.isReady) {
    return <h2>Loading...</h2>;
  }
```

**Comment by PReichetanz** [comment]
@actyralf @djfarly
This problem is not yet solved, could one of you has a look into that?

**Comment by actyralf** [comment]
I can't reproduce the error - what exactly is the problem? I can see a menu list and no errors besides a "missing key" warning.

**Reply by LuYujin9** [reply]
Please refresh the page. then an error will be there : "Error: Rendered more hooks than during the previous render."
My description was not so good, so I updated my question.

**Comment by actyralf** [comment, marked as answer]
The error occurs because you are calling `useEffect()` conditionally: It will be called in some cases (if `router.isReady` is true) and in others not (if `router.isReady` is false the function returns before calling useEffect()).
React relies on all hooks being called on every render, so this produces the error.
There is no need for a `useEffect()` call though, please try this code:

```js
export default function Menu() {
  const [index, setIndex] = useState(0);

  const router = useRouter();

  const { id } = router.query;
  if (!id) return;

  const restaurant = restaurants.find((restaurant) => restaurant.id === id);
  const foodsInCurrentPage = restaurant?.foods?.slice(index, index + 5) || [];

  function handleToNextPage() {
    index < foods.length - 5 ? setIndex(index + 5) : null;
  }
...
```

I renamed `FoodsInCurrentPage` to be camel case, so please adjust it in your `return` section accordingly:

```js
 <StyledList role="list">
          {foodsInCurrentPage.map((food) => (
            <StyledListItem key={food.id}>
              <FoodCard food={food} />
            </StyledListItem>
          ))}
```

**Reply by LuYujin9** [reply]
😄 🎉  Weee! It works. Thank you very much.
